### PR TITLE
[SPARK-36125][PYTHON] Implement non-equality comparison operators between two Categoricals

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Implement non-equality comparison operators between two Categoricals.
Non-goal: supporting Scalar input will be a follow-up task.

### Why are the changes needed?
pandas supports non-equality comparisons between two Categoricals. We should follow that.

### Does this PR introduce _any_ user-facing change?
Yes. No `NotImplementedError` for `<`, `<=`, `>`, `>=` operators between two Categoricals. An example is shown as below:

From:
```py
>>> import pyspark.pandas as ps
>>> from pandas.api.types import CategoricalDtype
>>> psser = ps.Series([1, 2, 3]).astype(CategoricalDtype([3, 2, 1], ordered=True))
>>> other_psser = ps.Series([2, 1, 3]).astype(CategoricalDtype([3, 2, 1], ordered=True))
>>> with ps.option_context("compute.ops_on_diff_frames", True):
...     psser <= other_psser
... 
Traceback (most recent call last):
...
NotImplementedError: <= can not be applied to categoricals.
```

To:
```py
>>> import pyspark.pandas as ps
>>> from pandas.api.types import CategoricalDtype
>>> psser = ps.Series([1, 2, 3]).astype(CategoricalDtype([3, 2, 1], ordered=True))
>>> other_psser = ps.Series([2, 1, 3]).astype(CategoricalDtype([3, 2, 1], ordered=True))
>>> with ps.option_context("compute.ops_on_diff_frames", True):
...     psser <= other_psser
... 
0    False
1     True
2     True
dtype: bool
```
### How was this patch tested?
Unit tests.

Keyword:SPARK-35337

### CODE REVIEW

File: python/pyspark/pandas/data_type_ops/categorical_ops.py [hunk 1]
```diff
@@ -64,15 +66,28 @@ def astype(self, index_ops: IndexOpsLike, dtype: Union[str, type, Dtype]) -> Ind
...
+        return column_op(Column.__ge__)(left, right)
 
-    def gt(self, left: IndexOpsLike, right: Any) -> SeriesOrIndex:
-        raise NotImplementedError(">= can not be applied to %s." % self.pretty_name)
+
+def _non_equality_comparison_input_check(left: IndexOpsLike, right: Any) -> None:
+    if not left.dtype.ordered:
+        raise TypeError("Unordered Categoricals can only compare equality or not.")
+    if isinstance(right, IndexOpsMixin) and isinstance(right.dtype, CategoricalDtype):
+        if hash(left.dtype) != hash(right.dtype):
```
Comment (xinrong-databricks) [inline python/pyspark/pandas/data_type_ops/categorical_ops.py:90]: Followed pandas comparison [here](https://github.com/pandas-dev/pandas/blob/master/pandas/core/arrays/categorical.py#L144)

**Comment by xinrong-databricks** [comment]
ERROR [2.347s]: test_reuse_worker_of_parallelize_range (pyspark.tests.test_worker.WorkerReuseTest)

**Comment by xinrong-databricks** [comment]
CC @ueshin @HyukjinKwon @itholic

File: python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py [hunk 2]
```diff
@@ -44,6 +44,26 @@ def other_pser(self):
     def other_psser(self):
         return ps.from_pandas(self.other_pser)
 
+    @property
+    def ordered_pser(self):
+        return pd.Series([1, 2, 3]).astype(CategoricalDtype([3, 2, 1], ordered=True))
+
+    @property
+    def disordered_psser(self):
```
Comment (ueshin) [inline python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py:52]: nit: Sounds like this should be `unordered_psser`?

Comment (ueshin) [reply python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py:52]: btw, shall we put this below `ordered_psser`?

**Review by ueshin — APPROVED** [review]
Otherwise, LGTM.

File: python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py (hunk 2 above)
Comment (xinrong-databricks) [reply python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py]: Good point! Adjusted.

**Comment by ueshin** [comment]
Thanks! merging to master/3.2.

**Pull request closed without merging**